hibernate-core/src/main/java/org/hibernate/query/sql/internal/NativeQueryImpl.java

```diff
@@ -355,32 +354,43 @@ private void checkResultType(Class<R> resultType, ResultSetMapping resultSetMapp
 					}
 					break;
 				default:
-					// The return type has to be a class with an appropriate constructor, i.e. one whose parameter types match
-					// the types of the result builders. If none such constructor is found, throw an IAE
+					// The return type has to be a class with an appropriate constructor,
+					// i.e. one whose parameter types match the types of the result builders.
+					// If no such constructor is found, throw an IAE
 					if ( !validConstructorFoundForResultType( resultType, resultSetMapping ) ) {
-						throw new IllegalArgumentException( "The declared return type for a multi-valued result set mapping should be Object[], Map, List, or Tuple" );
+						throw new IllegalArgumentException(
+								"The return type for a multivalued result set mapping should be Object[], Map, List, or Tuple"
+								+ " or it must have an appropriate constructor"
+						);
 					}
 			}
 		}
 	}
 
 	private boolean validConstructorFoundForResultType(Class<R> resultType, ResultSetMapping resultSetMapping) {
-		// Only 1 constructor with the right number of parameters is allowed (see NativeQueryConstructorTransformer)
-		Constructor<?> constructor = resultType.getConstructors()[0];
-		if ( constructor.getParameterCount() != resultSetMapping.getNumberOfResultBuilders() ) {
-			return false;
-		}
-		final List<ResultBuilder> resultBuilders = resultSetMapping.getResultBuilders();
-		Class<?>[] paramTypes = constructor.getParameterTypes();
-		for ( int i = 0; i < resultBuilders.size(); i++ ) {
-			if (
-					resultBuilders.get( i ).getJavaType() != ( paramTypes[i].isPrimitive() ?
-							getDescriptorByPrimitiveType(paramTypes[i] ).getWrapperClass() :
-							paramTypes[i]) ) {
-				return false;
+		// TODO: Only one constructor with the right number of parameters is allowed
+		//       (see NativeQueryConstructorTransformer) so we should validate that
```

Comment:
Wasn't I doing just that in my original code?
I don't quite understand why you would instead iterate all constructors, if clearly only exactly one with the right amt. of params is allowed ... ?
I took the constructor iteration i originally had out when I saw there should only be one ...

Reply:
Well we have to skip over all the constructors with the wrong number of parameters until we get to one with the right number of parameters.

Reply:
IINM, that is not true; `NativeQueryConstructorTransformer`should allow only a constructor with the same number of params, else an `InstantiationException`would be thrown, so no need to iterate?

Reply:
Ok, I was mislead by the exception message ...
Maybe it's somewhat confusing since it's possible for a resultclass to have several constructors with the same number of params as needed, but with different param types, and then you'd still get that message?
I'm just annoyed by the fact I initially had a perfectly good implementation including constructor iteration, and then I took that out thinking it was not needed, only for the lead Hibernate guy to come in after me and slap me on the wrist by rewriting my code. Fucks up my self-esteem in so many ways ...

Reply:
> it's possible for a resultclass to have several constructors with the same number of params as needed, but with different param types, and then you'd still get that message?

Correct. The potentially-ambiguous case is where you have multiple constructors with the same number of parameters but different types.

>  Fucks up my self-esteem in so many ways 

I make mistakes all the time. I don't let it bother me.